### PR TITLE
Add the possibility to extract data from the domain in components

### PR DESCRIPTION
**Pull request by tabergma**
**Proposed changes**:
closes https://github.com/RasaHQ/rasa/issues/7271

Add method `extract_data_from_domain` to components. The method is called from `create`. We cannot update the `init` method as this would break the code for any custom components.

For now just the `RegexEntityExtractor` makes use of the domain. 

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by tabergma** [comment]
@degiz Can you please review this PR? Thanks.

File: rasa/nlu/classifiers/mitie_intent_classifier.py [hunk 1]
```diff
@@ -106,16 +114,24 @@ def load(
...
     def persist(self, file_name: Text, model_dir: Text) -> Dict[Text, Any]:
+        """Persist this component to disk for future loading.
+
+        Args:
+            file_name: The file name of the model.
+            model_dir: The directory to store the model to.
 
-        if self.clf:
+        Returns:
+            An optional dictionary with any information about the stored model.
```
Comment (degiz) [inline rasa/nlu/classifiers/mitie_intent_classifier.py:129]: Why is it `optional`?

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 2]
```diff
@@ -204,22 +211,23 @@ def process(self, message: Message, **kwargs: Any) -> None:
...
-        Return probabilities for all labels.
-
-        :param X: bow of input text
-        :return: vector of probabilities containing one entry for each label"""
+        Args:
+            X: bow of input text
 
-        return self.clf.predict_proba(X)
+        Returns:
+            Vector of probabilities containing one entry for each label
```
Comment (degiz) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:218]:
```suggestion
            Vector of probabilities containing one entry for each label.
```

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 3]
```diff
@@ -204,22 +211,23 @@ def process(self, message: Message, **kwargs: Any) -> None:
     def predict_prob(self, X: np.ndarray) -> np.ndarray:
         """Given a bow vector of an input text, predict the intent label.
 
-        Return probabilities for all labels.
-
-        :param X: bow of input text
-        :return: vector of probabilities containing one entry for each label"""
+        Args:
+            X: bow of input text
```
Comment (degiz) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:215]:
```suggestion
            X: bow of input text.
```

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 4]
```diff
@@ -51,37 +51,46 @@ def required_components(cls) -> List[Type[Component]]:
...
     def required_packages(cls) -> List[Text]:
+        """Specifies which python packages need to be installed."""
         return ["sklearn"]
 
     def transform_labels_str2num(self, labels: List[Text]) -> np.ndarray:
         """Transforms a list of strings into numeric label representation.
 
-        :param labels: List of labels to convert to numeric representation"""
+        Args:
+            labels: List of labels to convert to numeric representation
```
Comment (degiz) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:77]:
```suggestion
            labels: List of labels to convert to numeric representation.
```

File: rasa/nlu/classifiers/mitie_intent_classifier.py [hunk 5]
```diff
@@ -23,16 +23,23 @@ def required_components(cls) -> List[Type[Component]]:
...
-        self, component_config: Optional[Dict[Text, Any]] = None, clf=None
+        self,
+        component_config: Optional[Dict[Text, Any]] = None,
+        classifier: Optional[Any] = None,
     ) -> None:
-        """Construct a new intent classifier using the MITIE framework."""
+        """Construct a new intent classifier using the MITIE framework.
 
+        Args:
+            component_config: the component configuration
```
Comment (degiz) [inline rasa/nlu/classifiers/mitie_intent_classifier.py:33]:
```suggestion
            component_config: The component configuration.
```

File: rasa/nlu/classifiers/mitie_intent_classifier.py [hunk 6]
```diff
@@ -23,16 +23,23 @@ def required_components(cls) -> List[Type[Component]]:
...
+            classifier: the MITIE classifier
```
Comment (degiz) [inline rasa/nlu/classifiers/mitie_intent_classifier.py:34]:
```suggestion
            classifier: The MITIE classifier.
```

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 7]
```diff
@@ -204,22 +211,23 @@ def process(self, message: Message, **kwargs: Any) -> None:
...
     def predict(self, X: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
         """Given a bow vector of an input text, predict most probable label.
 
-        Return only the most likely label.
-
-        :param X: bow of input text
-        :return: tuple of first, the most probable label and second,
-                 its probability."""
+        Args:
+            X: bow vector of input text
```
Comment (degiz) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:226]:
```suggestion
            X: Bow vector of input text.
```

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 8]
```diff
@@ -204,22 +211,23 @@ def process(self, message: Message, **kwargs: Any) -> None:
...
 
+        Returns:
+            Tuple of the most probable label and its probability.
```
Comment (degiz) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:229]:
```suggestion
            The most probable label and its probability.
```

File: rasa/nlu/components.py [hunk 9]
```diff
@@ -879,11 +906,18 @@ def create_component(
...
 
-    def create_component_from_class(self, component_class: Type[C], **cfg: Any) -> C:
+    def create_component_from_class(self, component_class: Type[C], **config: Any) -> C:
         """Create a component based on a class and a configuration.
 
-        Mainly used to make use of caching when instantiating component classes."""
+        Mainly used to make use of caching when instantiating component classes.
 
+        Args:
+            component_class: the component class
```
Comment (degiz) [inline rasa/nlu/components.py:915]:
```suggestion
            component_class: The component class.
```

File: rasa/nlu/components.py [hunk 10]
```diff
@@ -879,11 +906,18 @@ def create_component(
...
+            **config: the configuration
```
Comment (degiz) [inline rasa/nlu/components.py:916]:
```suggestion
            **config: The configuration.
```

Comment (degiz) [reply rasa/nlu/components.py:916]: Could we also specify which `configuration` that is?

File: rasa/nlu/utils/pattern_utils.py [hunk 11]
```diff
@@ -1,25 +1,26 @@
...
-    training_data: TrainingData, use_only_entities: bool = False
+    training_data: TrainingData, pattern_names: Optional[List[Text]] = None
 ) -> List[Dict[Text, Text]]:
     """Convert the lookup tables from the training data to regex patterns.
+
     Args:
         training_data: The training data.
-        use_only_entities: If True only regex features with a name equal to a entity
-          are considered.
+        pattern_names: List of pattern names to use. If list is empty or None all
```
Comment (degiz) [inline rasa/nlu/utils/pattern_utils.py:15]:
```suggestion
        pattern_names: List of pattern names to use. If list is empty or `None` all
```

File: rasa/nlu/utils/pattern_utils.py [hunk 12]
```diff
@@ -116,8 +117,8 @@ def extract_patterns(
 
     Args:
         training_data: The training data.
-        use_only_entities: If True only lookup tables and regex features with a name
-          equal to a entity are considered.
+        patter_names: List of pattern names to use. If list is empty or None all
```
Comment (degiz) [inline rasa/nlu/utils/pattern_utils.py:120]:
```suggestion
        patter_names: List of pattern names to use. If list is empty or `None` all
```

File: rasa/nlu/components.py [hunk 13]
```diff
@@ -754,6 +765,17 @@ def can_handle_language(cls, language: Hashable) -> bool:
         else:
             return language not in not_supported_language_list
 
+    def extract_data_from_domain(self, domain: Domain) -> None:
```
Comment (degiz) [inline rasa/nlu/components.py:768]: Thinking out loud: do we actually need this method to be in the base class aka interface? 

We don't control the `create()` method. Users (and even we ourselves, see `MitieNLP`) overwrite it, so nothing guaranties that `extract_data_from_domain()` will be called (as opposed to other public methods of `Component`, which are called by ROS code). That's why IMO the `extract_data_from_domain()` doesn't belong to the interface.

Also, currently the only implementation says that `Extract the names of all entities from the given domain`, so the better name inside this component would be `extract_entity_names_from_domain`.

I'd rather implement some "helper extract" method only in those implementations that actually need it.

What do you think?

**Review by degiz — COMMENTED** [review]
Great job adding the docstrings and renaming properties! 👍 

I've added few comments about the docstrings, and a comment about the need of `extract_data_from_domain` method, I'd love to hear your opinion!

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 14]
```diff
@@ -204,22 +211,23 @@ def process(self, message: Message, **kwargs: Any) -> None:
     def predict_prob(self, X: np.ndarray) -> np.ndarray:
         """Given a bow vector of an input text, predict the intent label.
 
-        Return probabilities for all labels.
-
-        :param X: bow of input text
-        :return: vector of probabilities containing one entry for each label"""
+        Args:
+            X: bow of input text.
```
Comment (degiz) [inline rasa/nlu/classifiers/sklearn_intent_classifier.py:215]:
```suggestion
            X: Bow of input text.
```

**Review by degiz — APPROVED** [review]
Looks great 👍

**Pull request merged**